### PR TITLE
Add support for interleaved moe

### PR DESCRIPTION
**Pull request by Macchiato123000**
Our team is working on interleaved moe model architecture (dense llama layer and mixture of expert layer interleaved layer by layer), and TensorRT-LLM mainline code doesn't support interleaved moe yet, so we made local changes to support it. In our huggingface config.json, we have attribute `moe_layers` available to specify which layers are moe layers in the range of num_layers, the rest will be dense llama layers.

Also in our dense layers, the mlp weights are named as `mlp.w3.weight`, `mlp.w2.weight` and `mlp.w1.weight` rather than `mlp.up_proj.weight`, `mlp.down_proj.weight`, `mlp.gate.weight`, so adding support (backwards-compatibly) for `mlp.w3.weight`, `mlp.w2.weight` and `mlp.w1.weight`.

### CODE REVIEW

**Comment by nv-guomingz** [comment]
@Macchiato123000 thanks for your contribution.
Is there a specific model that we could download to verify the MR's correctness?

File: tensorrt_llm/models/llama/model.py [hunk 1]
```diff
@@ -65,11 +65,16 @@ def __init__(self, config: LLaMAConfig, layer_idx: int):
             tp_rank=config.mapping.tp_rank,
             quant_mode=config.quant_mode)
 
-        mlp_hidden_size = config.hidden_size * 4 if config.intermediate_size is None else config.intermediate_size
+        if int(self.local_layer_idx) % 2 == 0:
```
Comment (yuxianq) [inline tensorrt_llm/models/llama/model.py]: We can check layer_idx in moe_layers instead of assuming that moe_layers is perfectly interleaved.

Comment (Macchiato123000) [reply tensorrt_llm/models/llama/model.py]: Good call, updated in the new commit for this place and all related.

**Comment by yuxianq** [comment]
@Macchiato123000 Can you provide a Hugging Face model to validate this PR? I think the LlamaForCausalLM architecture does not support interleaved moe, is your model a new architecture?

**Comment by Macchiato123000** [comment]
> @Macchiato123000 Can you provide a Hugging Face model to validate this PR? I think the LlamaForCausalLM architecture does not support interleaved moe, is your model a new architecture?

Yes it's a new architecture that our team is working on.
It's like https://huggingface.co/papers/2401.04081 moe-mambda each layer by the other, but the dense layers are Llama instead. I didn't find anything alike in huggingface. 
I'm trying to upload and share a model sample checkpoint to s3 and grant public access (github doesn't allow large file upload unfortunately)

**Comment by Macchiato123000** [comment]
> > @Macchiato123000 Can you provide a Hugging Face model to validate this PR? I think the LlamaForCausalLM architecture does not support interleaved moe, is your model a new architecture?
> 
> Yes it's a new architecture that our team is working on. It's like https://huggingface.co/papers/2401.04081 moe-mambda each layer by the other, but the dense layers are Llama instead. I didn't find anything alike in huggingface. I'm trying to upload and share a model sample checkpoint to s3 and grant public access (github doesn't allow large file upload unfortunately)

Here's the sample model ckpt to verify the code - https://ckt-public-share.s3.amazonaws.com/m16x2b_exp4-5_280k_safetensors_toy.zip
btw can view `config.json` file [here](https://ckt-public-share.s3.amazonaws.com/m16x2b_exp4-5_280k_safetensors_toy/config.json) and `model.safetensors.index.json` [here](https://ckt-public-share.s3.amazonaws.com/m16x2b_exp4-5_280k_safetensors_toy/model.safetensors.index.json)

I verified by running the below commands locally:
```
python3 convert_checkpoint.py --model_dir /opt/checkpoints/m16x2b_exp4-5_280k_safetensors_toy/ \
                            --output_dir /opt/converted_checkpoint/m16x2b_exp4-5_280k_safetensors_toy/tllm_checkpoint_8gpu_tp8 \
                            --dtype float16 \
                            --tp_size 4 \
                            --moe_tp_size 4 \
                            --moe_ep_size 1 \
                            --workers 4 
```

```
trtllm-build \
--checkpoint_dir /opt/converted_checkpoint/m16x2b_exp4-5_280k_safetensors_toy/tllm_checkpoint_8gpu_tp8 \
--output_dir /opt/trt_llm/examples/mixtral/trt_engines/mixtral/converted_checkpoint/m16x2b_exp4-5_280k_safetensors_toy/tllm_checkpoint_8gpu_tp8/bs16 \
--max_input_len 1024 \
--max_seq_len 1152 \
--max_batch_size 16 \
--gpt_attention_plugin float16 \
--gemm_plugin float16 \
--use_paged_context_fmha enable \
--paged_kv_cache enable \
--max_num_tokens 128 \
--workers 4
```

```
mpirun -np 4 --allow-run-as-root \
python3 /opt/trt_llm/examples/run.py \
--engine_dir /opt/trt_llm/examples/mixtral/trt_engines/mixtral/converted_checkpoint/m16x2b_exp4-5_280k_safetensors_toy/tllm_checkpoint_8gpu_tp8/bs16 \
--tokenizer_dir /opt/tokenizers/godzilla_tokenizer \
--max_output_len=128 \
--max_input_length=128 \
--input_file /opt/trt_llm/numbers.txt \
--run_profiling
```

**Comment by Macchiato123000** [comment]
> @Macchiato123000 thanks for your contribution. Is there a specific model that we could download to verify the MR's correctness?

There's no public huggingface model that I'm aware of, but we have this sample checkpoints for verification.

https://ckt-public-share.s3.amazonaws.com/m16x2b_exp4-5_280k_safetensors_toy.zip
btw can view `config.json` file [here](https://ckt-public-share.s3.amazonaws.com/m16x2b_exp4-5_280k_safetensors_toy/config.json) and `model.safetensors.index.json` [here](https://ckt-public-share.s3.amazonaws.com/m16x2b_exp4-5_280k_safetensors_toy/model.safetensors.index.json)

**Comment by yuxianq** [comment]
@Macchiato123000 You use `MixtralForCausalLM` as architecture, but the `modeling_mixtral.py` in transformers cannot support `moe_layers`. You need to use a new architecture and create a PR to transformers. After transformers's PR merged, we can continue to integrate your PR to TRT-LLM. Before that, you can hold your updates in this side branch. TRT-LLM dose not prefer to support a private model officially.

**Comment by Macchiato123000** [comment]
> @Macchiato123000 You use `MixtralForCausalLM` as architecture, but the `modeling_mixtral.py` in transformers cannot support `moe_layers`. You need to use a new architecture and create a PR to transformers. After transformers's PR merged, we can continue to integrate your PR to TRT-LLM. Before that, you can hold your updates in this side branch. TRT-LLM dose not prefer to support a private model officially.

Hi, I believe it takes the code path from TensorRT-LLM package itself from [here](https://github.com/NVIDIA/TensorRT-LLM/blob/main/examples/llama/convert_checkpoint.py#L415) which calls [LlamaForCausalLM.from_hugging_face](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/models/llama/model.py#L285,L309) instead of directly calling `modeling_mixtral.py` in transformers.

And we don't have a plan to share our private model yet to transformers though.

**Comment by yuxianq** [comment]
> I believe it takes the code path from TensorRT-LLM package itself from [here](https://github.com/NVIDIA/TensorRT-LLM/blob/main/examples/llama/convert_checkpoint.py#L415) which calls [LlamaForCausalLM.from_hugging_face](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/models/llama/model.py#L285,L309) instead of directly calling modeling_mixtral.py in transformers.

Yes, but the model spec comes from modeling_mixtral.py. We cannot say a model is `MixtralForCausalLM` if its structure is different from the definition in modeling_mixtral.py. You need to update modeling_mixtral.py to make your model a "mixtral". transformers may not allow you to update an existed model, so push a new architecture to transformers may be a better solution.

**Comment by hchings** [comment]
Hi @Macchiato123000, we're closing this out as TensorRT-LLM in general does not support private model. 
Please feel free to reach out to us if you have any questions to support this private model at your end.

**Pull request closed without merging**